### ISSUE TITLE
Update Citra from upstream

### ISSUE THREAD

**Issue by RobLoach** [libretro/citra#59]
Last update was September 30th, 2018. Would be good to update Citra!

**Comment by mikeguidry** [comment]
I am attempting to use libretro_citra.cpp, etc directly on top of the current citra code.. It has a few changes which allow gateway cheats, etc.  I moved all of the changes I could find (Cmake, libretro/emuwindow, etc) and I made a few changes to get it to compile.  It was simple things like OpenGL::, . the memory thing that was already discussed (new version goes from Kernel::memory_regions[0].linear_heap_memory->data(); to class Impl.., and a few other small things...

Anyway, It compiles fine, although has some issue with GL functions being NULL during renderer_opengl.cpp line 547.. so I attempted to hack a bit (force open gl to load functions in retro_run() since i assumed something changed with retro_reset not being called before load_game()).... I was able t get the functions to resolve, athough it returned a NULL string due to no GL context.  I am working to figure out where it should synchronize.  Most new updated documents, etc use this glsm.h, although citra is based on glad.h... so I am now researching some more, and came across this WIP. anyways..  I am trying to track it down.. it's my first time playing with any GL functions.

I do have a question though.  All of the commits relate to libretro specific functions from what I can see.  I am not sure if I somehow missed something.  I am wondering why your choosing to pull certain commits from upstream, rather than starting a new branch directly from the most recent version.  I noticed that the base, or what has been pulled in thus far doesn't have the architecture modifications which lead to Cheats via Gateway codes...I assumed I could just get it to function, and pair the retro cheat functions with citras.. 

Are there more modifications which I am not seeing that cause the project to attempt to merge commits rather than migrate the libretro code (as it seems separate to me) into the current version?

**Comment by j-selby** [comment]
@mikeguidry that isn't really going to be a workable solution, unfortunately - loading symbols each retro_run is going to be bad news. The core already handles symbol loading, and you need the entire citra_libretro stack for it all to work.

**Comment by mikeguidry** [comment]
but I did copy the citra_libretro 'stack...' at least all files, etc.  Were there direct source code modifications to citra only files? (not in libretro directory?) .. that's sort of what i was asking overall i guess :)

**Comment by mikeguidry** [comment]
I had traced back all commit logs from the initial release.. and all files modified I had moved over in my diff.. and attempted..I understand that reloading the func pointers isnt smart for every retro_run() which is loop() type.. I maybe didn't express correctly.. I meant context_reset().. Anyways, something changed when I moved all files i THOUGHT were necessary into the new source.  Unless I am missing something somewhere, I am unsure why the GL context is empty (Null pointers) and when I force loaded it.. then the GlGetString(*VERSION*) scenario brought back a NULL string.. No big deal.. I only spent 3 hours on it.. I'll attempt again but I figured I'd ask this evening.

I didn't notice anything else*important* unless theres some structural type change in code not located in libretro_citra... or it could even be thewrong header file, or im using ubuntu boost library rather than the travis.yml to compie... these are the next things I'm going to attempt since I had to make the one hack for -fPIC... anyways no big deal if nobody knows lol I  I'l post when I get further ;)

**Comment by mikeguidry** [comment]
Alright.. It took a bit longer before I was able to get back to it but anyhow.. I was able to get it running without crashing. The EmuWindow_LibRetro::ShouldDeferRendererInit() dissapeared, and it took awhile to track it down.  I had to learn abunch of glad/reasons for GL not initializing, etc..... Anyways...
Now I'm at a black screen, I see a cross hair (maybe mouse tracker) printing, and I can move mouse for 3-4 seconds, and then it stops updating the screen.  The sound is also not complete, as I was going to work out the differences in the audio sink after this.  Anyhow.. I do think its extremely plausible to work from this direction with your libretro code to obtain newest code + cheats quickly. I had to change something related to shaders, and versions as well.

Black screen or not.. if I click around it attempts to load my save game on Zelda.  This shows me that it is properly executing.. I just have to find the API to connect the backend opengl with the libretro buffer to get it completely working.

I'll get back to this soon to finish up... I can redo all changes again directly from when I copy the files to have a better history.

https://github.com/mikeguidry/citra/tree/libretro

**Comment by rharish101** [comment]
Any progress? It's been a year, and issues like #64 might start piling up.

**Comment by ordovice** [comment]
I'd like to add a +1 for updating this from upstream.

**Comment by Shoegzer** [comment]
Yes definitely would be quite appreciated, especially now with [so many new features and enhancements](https://citra-emu.org/entry/citra-progress-report-2020-q2/) being added upstream.

**Comment by Sanic1239** [comment]
I have no idea how much effort it takes to port emulators to Libretro, but would updating this to upstream cause any problems? I'm just wondering why not all Libretro cores are updated to upstream.

**Comment by budtz** [comment]
Maybe theres so many changes that it’s a very big job to update it?
